### PR TITLE
Refactoring: centralize loop over test and introduce classes to defined decks.

### PR DESCRIPTION
**Pull request by CusiniM**
GEOS PR: https://github.com/GEOS-DEV/GEOS/pull/2624 

In this PR:

- we centralize the loop over over the decks in a unique location.
- introduce classes to define the tests

It's far from being perfect and it's only a first step and we can make those objects evolve so that we can have more flexibility and introduce new tests (e.g., comparison of solution between parallel and serial runs) but it seemed an important first step. 

Unfortunately, some tests had naming conventions a bit different from others so this PR will require creating new baselines and deleting some old ones. 

Btw, if we think the ownership of the tests is important, I can easily add it back.

### CODE REVIEW

**Review by TotoGaz — APPROVED** [review]

File: scripts/geos_ats_package/geos_ats/test_builder.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    
+    @classmethod
+    def from_dict(cls, data_dict):
+        return cls(
+            name=data_dict["name"],
+            description=data_dict["description"],
+            partitions=data_dict["partitions"],
+            restart_step=data_dict["restart_step"],
+            check_step=data_dict["check_step"]
+        )
```
[the remark this answers is not in the document]
Comment (CusiniM) [reply scripts/geos_ats_package/geos_ats/test_builder.py]: actually, I have tried but then the original dict must have exactly the same fields and in some cases the dict contains more stuff than what's in TestDeck. I think eventually we will be able to get rid of this method though.

File: scripts/geos_ats_package/geos_ats/test_builder.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    curves: list[list[str]]
+
+    def as_dict(self):
+        return asdict(self)
+
+@dataclass
+class TestDeck:
+    name: str
+    description: str
+    partitions: list[tuple[int, int, int]]
```
Comment (TotoGaz) [inline scripts/geos_ats_package/geos_ats/test_builder.py:31]:
```suggestion
    partitions: Iterable[tuple[int, int, int]]
```

File: scripts/geos_ats_package/geos_ats/test_builder.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+class CurveCheckParameters:
+    filename: str
+    tolerance: list[float]
+    script_instructions: list[list[str]]
+    curves: list[list[str]]
+
+    def as_dict(self):
+        return asdict(self)
+
+@dataclass
```
Comment (TotoGaz) [inline scripts/geos_ats_package/geos_ats/test_builder.py:27]: What about using [frozen instances](https://docs.python.org/3/library/dataclasses.html#frozen-instances)?

File: tests/allTests/sedov/sedov.ats [hunk 4]
```diff
@@ -10,36 +8,19 @@ curvecheck_params['tolerance'] = 1e-10
...
-    testcase_name = "sedov_%d" % N
+sedov = {
+    "name": "sedov",
+    "description": "Test the basic sedov problem and restart capabilities",
+    "partitions": partitions,
+    "restart_step": 50,
+    "check_step": 100
+}
+
+decks = ( sedov, )
```
Comment (TotoGaz) [inline tests/allTests/sedov/sedov.ats:21]: I think that you can create a `TestDeck` instance here.

File: tests/allTests/solidMechanicsSSLE/SSLE.ats [hunk 5]
```diff
@@ -1,35 +1,21 @@
...
-                         restartcheck_params=restartcheck_params)))
+sedovSSLE = {
+    "name": "SSLE-sedov",
+    "description": "Test the small strain linear elastic solver",
+    "partitions": partitions,
+    "restart_step": 50,
+    "check_step": 100
+}
+
+decks = ( sedovSSLE, )
```
Comment (TotoGaz) [inline tests/allTests/solidMechanicsSSLE/SSLE.ats:19]: You can create a `TestDeck` instance here as well

File: tests/allTests/wavePropagation/ElasticSEM.ats [hunk 6]
```diff
@@ -1,52 +1,33 @@
...
+                  partitions=((1, 1, 1), (2, 2, 2)), 
+                  restart_step=100, 
+                  check_step=200,
+                  restartcheck_params=RestartcheckParameters(**restartcheck_params)), 
+         TestDeck(name='elas3D_Q3_abc_smoke', 
+                  description='Elastic wave solver, third-order FE, absorbing BC', 
+                  partitions=((1, 1, 1), (2, 2, 2)), restart_step=100, 
+                  check_step=200), 
+                  TestDeck(name='elas3D_Q3_abc_fs_smoke', 
+                           description='Elastic wave solver, third-order FE, absorbing BC on the sides, free surface BC at the top',
```
Comment (TotoGaz) [inline tests/allTests/wavePropagation/ElasticSEM.ats:27]: Indentation?

File: tests/allTests/solidMechanicsSSLE/plasticCubeReset.ats [hunk 7]
```diff
@@ -1,35 +1,22 @@
...
-                         restartcheck_params=restartcheck_params)))
+plasticCubeReset = {
+    "name": "plasticCubeReset",
+    "description": "Test the initialization step of for solid mechanics",
+    "partitions": partitions,
+    "restart_step": 4,
+    "check_step": 7
+}
+
+decks = ( plasticCubeReset, )
```
Comment (TotoGaz) [inline tests/allTests/solidMechanicsSSLE/plasticCubeReset.ats:19]: `TestDeck` instance?

File: tests/allTests/solidMechanicsSSLE/gravity.ats [hunk 8]
```diff
@@ -5,31 +5,18 @@ restartcheck_params["atol"] = 1.0E-5
...
-                         restartcheck_params=restartcheck_params)))
+gravity = {
+    "name": "gravity",
+    "description": "Test the gravity application in solid mechanics solver",
+    "partitions": partitions,
+    "restart_step": 1,
+    "check_step": 2
+}
+
+decks = ( gravity, )
```
Comment (TotoGaz) [inline tests/allTests/solidMechanicsSSLE/gravity.ats:17]: `TestDeck` instance?

File: tests/allTests/solidMechanicsSSLE/beamBending.ats [hunk 9]
```diff
@@ -8,54 +8,39 @@ curvecheck_params['tolerance'] = 0.0002
...
+    "name": "beamBending_hybridHexPrism_smoke",
+    "description": "Tests beam bending on general polyhedral mesh",
+    "partitions": partitions,
+    "restart_step": 0,
+    "check_step": 1
+}
+
+decks = ( beamBending,
+          beamBending_vem,
+          beamBending_hybridHexPrism )
```
Comment (TotoGaz) [inline tests/allTests/solidMechanicsSSLE/beamBending.ats:40]: `TestDeck` instance?

**Review by TotoGaz — APPROVED** [review]
I think we're very close from being able to remove the `from_dict` and use `TestDeck` instances everywhere?

File: scripts/geos_ats_package/geos_ats/test_builder.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    for deck in decks:
+            
+        deck_instance = TestDeck.from_dict( deck )
+        deck_instance.restartcheck_params = restartcheck_params_instance
+
+        if curvecheck_params_base is not None:
+            if ("aperture_curve_method" in deck) and ("pressure_curve_method" in deck):
+                curvecheck_params = curvecheck_params_base.copy()
+                curvecheck_params["script_instructions"][0][1] = deck["aperture_curve_method"]
+                curvecheck_params["script_instructions"][1][1] = deck["pressure_curve_method"]
```
Comment (cssherman) [inline scripts/geos_ats_package/geos_ats/test_builder.py:116]: This check is specific to Hydrofracture.ats.  It was originally designed for us to have a common test definition, and update the script instruction method name in the primary loop ('kgdToughnessDominated', etc.)

Comment (cssherman) [reply scripts/geos_ats_package/geos_ats/test_builder.py:116]: Looking at the Hydrofracture ats script, it seems like you have already expanded this.  So this should be able to be deleted

**Review by cssherman — APPROVED** [review]

**Review by rrsettgast — APPROVED** [review]

File: scripts/geos_ats_package/geos_ats/test_builder.py (hunk 3 above)
Comment (CusiniM) [reply scripts/geos_ats_package/geos_ats/test_builder.py:27]: We could. Although I don't love the fact that it does not allow for mixed behavior.

File: scripts/geos_ats_package/geos_ats/test_builder.py (hunk 10 above)
Comment (CusiniM) [reply scripts/geos_ats_package/geos_ats/test_builder.py:116]: No, I have not unfortunately, but it's the only one left. I could remove the dictionaries from there as well.

**Pull request merged**